File: solutions/python/all-your-base/1/all_your_base.py
from math import log


def rebase(input_base, digits, output_base):
    if input_base < 2:
        raise ValueError("input base must be >= 2")
    if output_base < 2:
        raise ValueError("output base must be >= 2")
    if any(not (0 <= d < input_base) for d in digits):
        raise ValueError("all digits must satisfy 0 <= d < input base")

    base_10 = sum(n * (input_base**b) for n, b in zip(reversed(digits), range(len(digits))))

    if base_10 == 0:
        return [0]

    ans = []

    for exp in range(int(log(base_10) / log(output_base)), -1, -1):
        ans.append(base_10 // (output_base**exp))
        base_10 %= output_base**exp

    return ans

File: solutions/go/all-your-base/1/all_your_base.go
package allyourbase

import (
	"errors"
	"math"
)

// ConverToBase converts the given input, seperate into each digit, into a different base.
func ConvertToBase(inputBase int, inputDigits []int, outputBase int) ([]int, error) {
	if inputBase < 2 {
		return nil, errors.New("input base must be >= 2")
	} else if outputBase < 2 {
		return nil, errors.New("output base must be >= 2")
	}

	outNum := 0

	for idx, num := range inputDigits {
		if num < 0 || num >= inputBase {
			return nil, errors.New("all digits must satisfy 0 <= d < input base")
		}
		outNum += num * int(math.Pow(float64(inputBase), float64(len(inputDigits)-idx-1)))
	}

	if outNum == 0 || len(inputDigits) == 0 {
		return []int{0}, nil
	}

	outputDigits := []int{}

	for exp := int(math.Log(float64(outNum)) / math.Log(float64(outputBase))); exp >= 0; exp-- {
		divisor := int(math.Pow(float64(outputBase), float64(exp)))
		digit := outNum / divisor
		outputDigits = append(outputDigits, digit)
		outNum %= divisor
	}

	return outputDigits, nil
}
